log/validator/validator.go

```diff
@@ -0,0 +1,171 @@
+					v.lineCounter.WithLabelValues(t.Filename, "invalid checksum length").Inc()
+				} else {
+					v.lineCounter.WithLabelValues(t.Filename, "bad").Inc()
+				}
+				select {
+				case <-outputLimiter.C:
+					v.log.Errf("%s: %s %q", t.Filename, err, line.Text)
+				default:
+				}
+			} else {
+				v.lineCounter.WithLabelValues(t.Filename, "ok").Inc()
+			}
+		}
+	}()
+
+	v.tailers = append(v.tailers, t)
+
+	return nil
+}
+
+// Shutdown should be called before process shutdown
+func (v *Validator) Shutdown() {
+	for _, t := range v.tailers {
+		// The tail module seems to have a race condition that will generate
+		// errors like this on shutdown:
+		// failed to stop tailing file: <filename>: Failed to detect creation of
+		// <filename>: inotify watcher has been closed
+		// This is probably related to the module's shutdown logic triggering the
+		// "reopen" code path for files that are removed and then recreated.
+		// These errors are harmless so we ignore them to allow clean shutdown.
+		_ = t.Stop()
+		t.Cleanup()
+	}
+}
+
+func lineValid(text string) error {
+	// Line format should match the following rsyslog omfile template:
+	//
+	//   template( name="LELogFormat" type="list" ) {
+	//  	property(name="timereported" dateFormat="rfc3339")
+	//  	constant(value=" ")
+	//  	property(name="hostname" field.delimiter="46" field.number="1")
+	//  	constant(value=" datacenter ")
+	//  	property(name="syslogseverity")
+	//  	constant(value=" ")
+	//  	property(name="syslogtag")
+	//  	property(name="msg" spifno1stsp="on" )
+	//  	property(name="msg" droplastlf="on" )
+	//  	constant(value="\n")
+	//   }
+	//
+	// This should result in a log line that looks like this:
+	//   timestamp hostname datacenter syslogseverity binary-name[pid]: checksum msg
+
+	fields := strings.Split(text, " ")
+	const errorPrefix = "log-validator:"
+	// Extract checksum from line
+	if len(fields) < 6 {
+		return fmt.Errorf("%s line doesn't match expected format", errorPrefix)
+	}
+	checksum := fields[5]
+	_, err := base64.RawURLEncoding.DecodeString(checksum)
+	if err != nil || len(checksum) != 7 {
+		return fmt.Errorf(
+			"%s expected a 7 character base64 raw URL decodable string, got %q: %w",
+			errorPrefix,
+			checksum,
+			errInvalidChecksum,
+		)
+	}
+
+	// Reconstruct just the message portion of the line
+	line := strings.Join(fields[6:], " ")
+
+	// If we are fed our own output, treat it as always valid. This
+	// prevents runaway scenarios where we generate ever-longer output.
+	if strings.Contains(text, errorPrefix) {
+		return nil
+	}
+	// Check the extracted checksum against the computed checksum
+	if computedChecksum := log.LogLineChecksum(line); checksum != computedChecksum {
+		return fmt.Errorf("%s invalid checksum (expected %q, got %q)", errorPrefix, computedChecksum, checksum)
+	}
+	return nil
+}
+
+// ValidateFile validates a single file and returns
+func ValidateFile(filename string) error {
+	file, err := os.ReadFile(filename)
```

Comment:
Consider reading the file line-by-line instead of loading it all into memory:

```go
file, err := os.Open(filename)
if err != nil {
    return err
}
defer file.Close()

scanner := bufio.NewScanner(file)
badFile := false
lineNum := 0

for scanner.Scan() {
    lineNum++
    line := scanner.Text()
...
```

Reply:
Oh, I almost forgot, the scanner has one part that's a little wonky:

```go
// Check for any errors that occurred during the scan.
if err := scanner.Err()
if err != nil {
    return fmt.Errorf("error reading file: %w", err)
}
```

Reply:
I actually have an idea to just get rid of this function completely: it’s not used at all during the regular operation of log-validator so is extra code duplication, but I didn’t want to make too many changes while moving things around.  This function is currently unmodified from the source.  So I’d like to just leave it as-is for now and eliminate it in the next patch in this series.